### ISSUE TITLE
mixed content on stats page  

### ISSUE THREAD

**Issue by vineettalwar** [YOURLS/YOURLS#2383]
- [x] Check you are using the [**LATEST** release](https://github.com/YOURLS/YOURLS/releases/latest) or the development branch.
- [x] Make sure you have [**SEARCHED** existing issues](https://github.com/YOURLS/YOURLS/issues?utf8=%E2%9C%93&q=is%3Aissue) first. 
- [x] Read the [**GUIDELINES**](https://github.com/YOURLS/YOURLS/blob/master/.github/CONTRIBUTING.md) linked in the yellow notice box above.
- [x] **COMPLETE** this issue template.

### Technical details regarding my environment

* YOURLS version: 1.7.2
* Plugins enabled: none 
* PHP version: 5.5.9
* Any other useful information depending on context (server OS & version, MySQL version, browser version, …)
OS: Ubuntu 14.04.5 LTS
nginx/1.4.6

### Bug description 
- Mixed Content: The page at 'https://domain.name/url+' was loaded over HTTPS, but requested an insecure image 'http://www.google.com/s2/favicons?domain=domain.name'. This content should also be served over HTTPS.
- A similar issue and similar server setup like https://github.com/YOURLS/YOURLS/pull/2167

**What is the current behavior?**
console throws mixed content on stats page 

This is a bug because…
- hardcoded url in functions-info.php:243 "http://www.google..."

**What is the expected or desired behavior?**
- Without hard coded protocol and using //link 

**Debugging summary**
-  I did debug the is_ssl as well, 
- var_dump($_SERVER['HTTPS']) is returns null, despite ssl certificate is set . Thus, is_ssl returns false.

**Comment by dgw** [comment]
`$_SERVER['HTTPS']` not being set properly is a server configuration issue. YOURLS matches protocols when it detects that it's running over HTTPS, so you'd probably be well served to find the configuration problem and fix it. That will prevent other mixed-content errors and such from both YOURLS and any other apps that check whether SSL is enabled.

**Comment by PopVeKind** [comment]
PS you did not say what you are trying to use https for? Example, some people use https for **only the admin** area but not the short URLs.  Other people want to use https for **both admin and short URLs**.  Most people use http for everything!

**Comment by LeoColomb** [comment]
> Most people use http for everything!

Not really, using HTTP without TLS in production is now highly discouraged.

**Comment by PopVeKind** [comment]
@LeoColomb - Hey Leo, this might make a very good point of discussion.  Let's start with a few quotes...

[ Google search for TLS](https://www.google.com/search?q=TLS)
> Transport Layer Security (TLS) – and its predecessor, Secure Sockets Layer (SSL), which is now prohibited from use by the Internet Engineering Task Force (IETF) – are cryptographic protocols that provide communications security over a computer network.

[LeoColomb (above) April 14, 2018](https://github.com/YOURLS/YOURLS/issues/2383#issuecomment-381317158)
> using HTTP without TLS in production is now highly discouraged.

Very interesting... **"Highly Discouraged"** by whom within the YOURLS community?  Do you highly discourage running a YOURLS production site on http Leo?  Are you, Leo, recommending we dump http and only support https with TLS?  I'm serious and after hearing from others, I might be persuaded to get behind that recommendation.

 I am working on three YOURLS projects, MultiSite, Full-Fledged Installer and removing Core Constants (Options) from config.pho.  Each of these would be far easier to build and maintain if we dumped support for http and used only https running TLS.

I have a hand full of production sites I plan to use this on that mate WordPress, YouTube and YOURLS for better branding.  The YOURLS sites will be MultiSite and the plan is for nginx, https and TLS.  Therefore, I am really vested in improving YOURLS.

[Recent ozh comment Dec 16, 2017](https://github.com/YOURLS/YOURLS/issues/1254#issuecomment-352122921)

> For the record, as of writing, 89.9% of YOURLS install are on http instead of https. The truth is: the vast majority of people just don't care. It's a personal URL shortener, no sensitive data transmitted, it just works, period.

Leo, are you willing to **_force_** 89.9% of YOURLS installed sites to switch from http to https and TLS?  That would make my projects easier and clean up some rather clumsy code.  Nonetheless, the real question is, are 89.9% of YOURLS sites "wrong" or **"Highly Discouraged"** by whom within the YOURLS community?

### This is interesting, others please comment?

**Comment by PopVeKind** [comment]
Is @ozh now **_"highly discouraging"_** people from using YOURLS under http?  Last December ozh seemed to indicate that YOURLS under http is what most of the community wanted.  Has there been any change since then?

**Comment by dgw** [comment]
Using plain HTTP is not "highly discouraged" by anyone associated with YOURLS. The whole web industry is moving toward encrypting everything, motivated by big players like Google giving preferential treatment to HTTPS sites.

Neither @LeoColomb nor @ozh is expressing anything but observations about the shift in the larger web community.

**Comment by PopVeKind** [comment]
Ha Ha Ha! So if I got it right @dgw You are saying that @LeoColomb comment, that is quoting my comment which is highly specific about the YOURLS community only. And even more specific in that my posts were trying to help the Original Poster with his nginx YOURLS install under https. That this is just a random thought or "observations about the shift in the larger web community."

### Wow!

But why did he quote my very YOURLS specific comment with such a random observation that does not help the OP at all?

Leo is a leader in the YOURLS community, so when he makes a blanket statement...

> using HTTP without TLS in production is now highly discouraged.

While quoting an extremely YOURLS specific comment, I think it is fair to say that YOURLS developers should take note and pay attention.

On the other hand, @ozh - who is at the very top of the YOURLS community, seems to be saying most of the YOURLS community (89.9% at least) doesn't care about HTTPS or TLS and run their production YOURLS servers on HTTP without TLS.

These are different and opposing positions coming from what arguably could be called the YOURLS community leadership, in posts that are specifically about YOURLS and the YOURLS community.  I think it deserves to be clarified.

Frankly, after studying the YOURLS code for about a year now, I can see where the HTTP/HTTPS code could stand some real improvements (perhaps a total rewrite for https!), that may even help with performance.  But there are more important things that should be worked on.  For example, **removing all core constants** (except DB) would be a worthy project.  Likewise, the YOURLS code could be greatly improved by **getting rid of the core options.** (they confuse Noobs, are not needed and are the source of all problem configurations). I have YOURLS demo sites running as a proof of concept, showing the advantages of doing these.

**Comment by LeoColomb** [comment]
@PopVeKind Please do not try to turn our respective comments in a conflicted position, it's not very kind.

@ozh may report usage statistic regarding YOURLS installation. Yes, maybe a lot of old YOURLS instances keep working on a raw HTTP session or are wrongly configured as HTTP only but now behind a TLS session. (BTW what @ozh said is a fact, not a argument to reject any HTTPS migration)
The thing is the web world is moving over a HTTP-with-TLS-only world, for a many reasons (good and bad ones).

You asked @vineettalwar (sorry for the spam 😅) why he's planning to use HTTPS everywhere, I simply answer you that he may want to follow current guidelines, and that the argument of what people do currently is not what is encouraged today from the whole web ecosystem. 😃

**Comment by PopVeKind** [comment]
@LeoColomb Thanks Leo for the clarification on your comment.  It may not have appeared that way to you but I was really asking vineettalwar if he planned to use HTTPS for both Admin area and Short URLs, or for the Admin Area Only?

About the **_whole web ecosystem_**, I tend to agree with your assessment.  In fact, as soon as I get the MultiSite Plugin running, I will be putting up a few YOURLS sites using nginx and HTTPS.  However, that move is due to our new shop policy and effects hundreds of domains, most do not run YOURLS.

> The thing is the web world is moving over a HTTP-with-TLS-only world, for a many reasons.

Do you have an example how to of this?  How to run YOURLS over HTTP-with-TLS?  Or how to run WordPress over HTTP-with-TLS?  I assume TLS will still need an SSL certificate and therefore cannot run on a shared hosting like https://www.pair.com/webhosting/shared.html offers. Likewise, TLS-over-HTTP would still require the delay of the security handshake, which is what some people object to.

I can see why people would want to run WordPress via HTTPS.  For example the visual security clues on the browser. Likewise, SEO experts have begun reporting Google is giving ranking preference to HTTPS.  However, a 301 driven YOURLS site would care little about Google ranking!  Anything on the browser would just be a flash till the Long URL was arrived at.

By the way, the prototype MultiSite Plugin allows parking of domains or protocols.  For example, if you had one site set up to service http://ni.ce/yourls it could be made to service **parked domains** (without additional redirects):
- http://ni.ce/yourls
- https://ni.ce/yourls
- http://www.ni.ce/yourls
- https://www.ni.ce/yourls
- http://ti.ny/yourls
- https://ti.ny/yourls
- http://www.ti.ny/yourls
- https://www.ti.ny/yourls

These share the same database and keywords. The above would all go to the same Long URL.

As well as completely different domains with their own database tables and keyword sets. These would go to independent Long URLs as if it was a completely different YOURLS site,

**Comment by vineettalwar** [comment]
HI, 
I have debugged as you suggested earlier by @PopVeKind 
Following are the findings.
- I have set one example page. with 
```
<php?
var_dump($_SERVER['HTTPS']);
```
- http://example.com/test-one.php -> redirected to https://example.com/test-one.php and value is returned as NULL 
- https://example.com/test-one.php -> returned NULL as well 
- Checked user/config.php -> 
- YOURLS_SITE is set to https://example.com 
- YOURLS_ADMIN_SSL is set to true
- I am using https 
- I have then debugged using phpinfo() on server. 
- Found following info  ->
- Under PHP variables
` _SERVER["HTTP_X_FORWARDED_PROTO"]  is set to https`
- Under HTTP Headers section
`X-Forwarded-Proto is set to https`
- Under apache environment section ..
`HTTP_X_FORWARDED_PROTO is set to https`
- Enabled SSL protocols are following:
`TLS1.0, TLS1.1, TLS1.2`
- Regarding the server config: Here nginx is set as a front end server, and apache as backend server. and nginx works as reverse proxy for apache and SSL is set via nginx.

**Comment by LeoColomb** [comment]
Ah! So you're "just" behind a proxy (nginx here).

A simple plugin can resolve this issue:

```php

<?php
/*
Plugin Name: HTTPS behind a proxy
Plugin URI: https://github.com/YOURLS/YOURLS/issues/2383
Description: This plugin add checks HTTPS when behind a proxy.
Version: 1.0
Author: YOURLS
Author URI: https://yourls.org/
*/

yourls_add_filter('is_ssl', 'proxy_is_secure_forwarded');
function proxy_is_secure_forwarded($is_ssl) {
	return $is_ssl 
		|| (isset($_SERVER['HTTP_X_FORWARDED_PROTO']) && "https" == $_SERVER['HTTP_X_FORWARDED_PROTO']);
}
```

**Comment by vineettalwar** [comment]
Thanks for helping to solve the issue.

**Comment by vineettalwar** [comment]
Nope, It doesn't work unfortunately.  - Page is stuck in redirect loop, after this. 

```
This page isn’t working
example.com redirected you too many times.
```

**Comment by vineettalwar** [comment]
So, this is how it is solved, documenting for others for reference. 

- culprit was yourls_admin_ssl define in config.php 
- As soon, as I removed it, redirect loop was gone. 
- Then, activated. above plugin, this fixes the issue.... 
- Both admin area and short urls works on ssl perfectly.

**Comment by PopVeKind** [comment]
@vineettalwar - Thanks Vineet for returning with your solution.

I was looking at that the other day and wondering about it.

See HTTPS Plugin Development #2391

YOURLS_SITE can be set to http:// or https:// and seems to affect the whole site ShortURLs and Admin
YOURLS_ADMIN_SSL seems to be if someone uses http:// in YOURLS_SITE (for ShortURLs but wants to use https:// for the Admin area only.

I wondered what would happen if both were set.  This may be the answer.

The HTTPS Plugin would use a pulldown menu for everything to make it less confusing.

An example of what I am talking about can be seen at:
http://a.fil.net/s/admin/db-config.php
Look at the HTTPS option near the bottom.
This uses neither YOURLS_SITE nor YOURLS_ADMIN_SSL and I think the pulldown menu is easier for people to get the right settings. Which do you think would be easier for you?
